zebROS_ws/src/controllers_2023/src/four_bar_controller.cpp

```diff
@@ -0,0 +1,461 @@
+//HPP CONTENTS INTO CPP
+#ifndef FOUR_BAR_CONTROLLER_2023
+#define FOUR_BAR_CONTROLLER_2023
+
+//added _2023 to the two lines above ^
+
+#include <ros/ros.h>
+#include <hardware_interface/joint_command_interface.h>
+#include <realtime_tools/realtime_buffer.h>
+#include <controller_interface/multi_interface_controller.h>
+#include <talon_controllers/talon_controller_interface.h> // "
+#include <pluginlib/class_list_macros.h> //to compile as a controller
+#include "controllers_2023_msgs/FourBarSrv.h"
+
+#include "ddynamic_reconfigure/ddynamic_reconfigure.h"
+
+namespace four_bar_controller_2023
+{
+
+class FourBarCommand_2023
+{
+    public:
+        FourBarCommand_2023()
+        {
+            angle_ = 0;
+        }
+        FourBarCommand_2023(double angle)
+        {
+            angle_ = angle;
+        }
+        double GetAngle() const
+        {
+            return angle_;
+        }
+
+
+    private:
+        double angle_; // radians.
+        // lower limit = 0 (all the way to the left), upper limit = pi (all the way to the right)
+};
+//this is the actual controller, so it stores all of the  update() functions and the actual handle from the joint interface
+//if it was only one type, controller_interface::Controller<TalonCommandInterface> here
+class FourBarController_2023 : public controller_interface::MultiInterfaceController<hardware_interface::TalonCommandInterface>
+
+{
+    public:
+        FourBarController_2023()
+        {
+        }
+
+        //should this be hardware_interface::TalonCommandInterface instead? What's the reason to import RobotHW then get CommandInterface from that instead of just importing TalonCommandIface?
+        //answer to my question: the TalonCommandInterface is passed in if it's not a multiInterfaceController, and just one kind of joint is made!
+        virtual bool init(hardware_interface::RobotHW *hw,
+                          ros::NodeHandle             &root_nh,
+                          ros::NodeHandle             &controller_nh) override;
+        virtual void starting(const ros::Time &time) override;
+        virtual void update(const ros::Time &time, const ros::Duration &period) override;
+        virtual void stopping(const ros::Time &time) override;
+
+        bool cmdService(controllers_2023_msgs::FourBarSrv::Request &req,
+                        controllers_2023_msgs::FourBarSrv::Response &res);
+
+        double angleFromX(double x, bool below) {
```

Comment:
Same here, make this private

Reply:
And mark it const (after the args), since it doesn't modify any class member vars